### PR TITLE
[pytree] Add optimized `tree_leaves` implementation

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #112324
* __->__ #112323
* #112327

pytree is used in many hot paths for dynamo tracing and in many cases we don't
care about the tree spec and just want the flattened list. This improves
`pytree.tree_leaves` to not construct the spec which gives a noticeable
performance improvement when multiplied by the many times it gets called
during tracing.

Concretely, I see a 2x speedup compared to `tree_flatten` in this benchmark:
```python
import torch.utils._pytree as pytree
%timeit pytree.tree_flatten([((100, 100), (100, 1)), dict(device="cuda")])[0]
%timeit pytree.tree_leaves([((100, 100), (100, 1)), dict(device="cuda")])
```

cc @zou3519

### CODE REVIEW

**Review by lezcano — APPROVED** [review]

**Comment by lezcano** [comment]
cc @XuehaiPan in case you want to implement this in optree as well

**Review by lezcano — APPROVED** [review]
~~LGTM, but it'd be good to add the optimisations from @XuehaiPan as well~~ Wrong pr

**Review by XuehaiPan — APPROVED** [review]

**Comment by XuehaiPan** [comment]
> in case you want to implement this in optree as well

@lezcano I'll benchmark the overhead of creating the `treespec` object on the C++ side. I think it might be marginal.

File: torch/utils/_pytree.py [hunk 1]
```diff
@@ -274,7 +274,19 @@ def tree_flatten(pytree: PyTree) -> Tuple[List[Any], TreeSpec]:
...
+    node_type = _get_node_type(pytree)
+    flatten_fn = SUPPORTED_NODES[node_type].flatten_fn
+    child_pytrees, context = flatten_fn(pytree)
+
+    # Recursively flatten the children
+    result : List[Any] = []
+    for child in child_pytrees:
+        result += tree_leaves(child)
+
+    return result
```
Comment (XuehaiPan) [inline torch/utils/_pytree.py:289]: I think we could have an even further optimized implementation using generators. I haven't profiled this yet. The implementation in this PR will create intermediate `list` objects in the recursive `tree_leaves` call.

File: torch/utils/_pytree.py [hunk 2]
```diff
@@ -274,7 +274,19 @@ def tree_flatten(pytree: PyTree) -> Tuple[List[Any], TreeSpec]:
...
+    if _is_leaf(pytree):
+        return [pytree]
+
+    node_type = _get_node_type(pytree)
+    flatten_fn = SUPPORTED_NODES[node_type].flatten_fn
+    child_pytrees, context = flatten_fn(pytree)
+
+    # Recursively flatten the children
+    result : List[Any] = []
+    for child in child_pytrees:
```
Comment (jon-chuang) [inline torch/utils/_pytree.py:286]: nit: maybe it's better to have an internal `_tree_flatten` function that optionally computes tree spec. Then wrap it in the public APIs.

Likewise for `tree_structure`

File: torch/utils/_pytree.py (hunk 1 above)
Comment (peterbell10) [reply torch/utils/_pytree.py:289]: Yeah you don't even need a generator, we can pass the result as an argument to a helper and append directly into the list.
```python
def tree_leaves_helper(pytree, result):
    if _isleaf(pytree):
        result.append(pytree)
    # [snip]
    for child in child_pytrees:
        tree_leaves_helper(child, result)
```

This way there are no intermediates at all. In my small benchmark it was only a 5% uplift, and I didn't want to diverge too far from the `tree_flatten` implementation so they would be easier to keep in sync. Could maybe change both though in a further PR.

File: torch/utils/_pytree.py (hunk 2 above)
Comment (peterbell10) [reply torch/utils/_pytree.py:333]: I tried this and every way I tried made `tree_leaves` slower and none made `tree_structure` significantly faster so I think this is a reasonable position.

**Pull request closed without merging**